### PR TITLE
refactor(core,react): contract types

### PR DESCRIPTION
**Pull request by tmm**
## Description

Refactors and simplifies contract types. Particularly remove verbose system for marking contract properties as optional in hooks and switch to pure inference.

## Additional Information

- [x] I read the [contributing docs](/wagmi-dev/wagmi/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

Your ENS/address: awkweb.eth

### CODE REVIEW

File: packages/core/src/actions/contracts/multicall.test.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ describe('multicall', () => {
   })
 
   it('default', async () => {
-    expect(await multicall({ contracts })).toMatchInlineSnapshot(`
+    await expect(multicall({ contracts })).resolves.toMatchInlineSnapshot(`
```
Comment (tmm) [inline packages/core/src/actions/contracts/multicall.test.ts:53]: not sure why i changed this test file like this, but was into the whole `await expect(…).resolves. toMatchInlineSnapshot(…)` pattern

File: packages/react/src/hooks/contracts/useContractInfiniteReads.test-d.ts [hunk 2]
```diff
@@ -1,74 +1,89 @@
...
 import { BigNumber } from 'ethers'
-import {
-  mlootContractConfig,
-  wagmigotchiContractConfig,
-} from 'packages/core/test'
+import { mlootContractConfig } from 'packages/core/test'
 import { assertType, describe, it } from 'vitest'
 
-import { useContractReads } from './useContractReads'
+import { useContractInfiniteReads } from './useContractInfiniteReads'
```
Comment (tmm) [inline packages/react/src/hooks/contracts/useContractInfiniteReads.test-d.ts:6]: This file was incorrectly named `useContractReads.test-d.ts` before

File: packages/react/src/hooks/contracts/useContractReads.test-d.ts [hunk 3]
```diff
@@ -1,89 +1,74 @@
-import type { InfiniteData } from '@tanstack/react-query'
```
Comment (tmm) [inline packages/react/src/hooks/contracts/useContractReads.test-d.ts:1]: This file was incorrectly named `useContractInfiniteReads.test-d.ts` before

File: packages/react/src/types/index.ts [hunk 4]
```diff
@@ -47,6 +47,25 @@ declare module 'ethers/lib/utils.js' {
   ): string
 }
 
+/**
+ * Makes {@link TKeys} optional in {@link TType} while preserving type inference.
+ */
+// s/o trpc (https://github.com/trpc/trpc/blob/main/packages/server/src/types.ts#L6)
+export type InferOptional<TType, TKeys extends keyof TType> = Partial<
```
Comment (jxom) [inline packages/react/src/types/index.ts:54]: nit: `PartialBy`? This feels like a variant of `Partial`, except, we are targetting provided `TKeys`.

**Review by jxom — APPROVED** [review]

**Pull request merged**